### DISCUSSION TITLE
Flatgeobuf in PostGIS with JSONB

### DISCUSSION THREAD

**Discussion by tordans** [flatgeobuf/flatgeobuf#342] [Q&A]
Hey! We where trying to use Flatgeobuf as an export format for the data we create with [osm2pgsql](https://osm2pgsql.org/doc/manual.html). A common way to use osm2pgsql is to put data in JsonB columns. 

However, we just notices …
- JsonB is not supported by the Flatgeobuf export in PostGIS ([its commented out](https://github.com/postgis/postgis/blame/master/postgis/flatgeobuf.c#L249-L260))
- [The ticket was closed](https://trac.osgeo.org/postgis/ticket/4997) and I could not find a follow up ticket in the open TODOs that are marked in the code
- I was not able to find anything on how to create a ticket or where those `TODO`s would be tracked (to get notified should anything happen at some point)

I was hoping someone in this community can maybe point me to the right place to learn more. Or maybe someone is able to create a Ticket in the PostGIS Track to give this visibility?

**Comment by bjornharrtell** [comment]
Hi @tordans. It's simply a matter of missing support/feature. When I implemented the direct support to import/export FlatGeobuf in PostGIS I couldn't figure out how jsonb datum conversion worked and it wasn't top priority for me at the time so I gave it up.

It's still not high on my todo list. Contribution or other incentives to implement are welcome.

**Reply by tordans** [reply]
@bjornharrtell is there a section in the documentation that explains which features are (not) supported? Reading the official PostGIS Docs https://postgis.net/docs/ST_AsFlatGeobuf.html it sounds like everything is there. And the linked webpage does provide any details either.

My comment at https://github.com/flatgeobuf/flatgeobuf/discussions/342#discussioncomment-11806223 suggests that there are not only issues with JsonB the issues I had with casting and NULL values might be separate issues(?).

**Comment by rush42** [comment]
We found a workaround which allows us to use `ST_AsFlatGeoBuf` for exporting our data. We first retrieve all keys which occur in the jsonb column and then select them in our subquery.  Here is the related [code](https://github.com/FixMyBerlin/atlas-app/blob/190be8711bc71d5313e8747c7044fcf9d1eaca6f/app/src/instrumentation/initExportFunctions.ts#L15-L44). 

It would be helpful to include a note in the documentation mentioning that jsonb support is not available at the moment.

**Reply by tordans** [reply]
The answer we hoped to have found in https://github.com/flatgeobuf/flatgeobuf/discussions/342#discussioncomment-10464221 is only a workaround at best.

Here are some issues with it:

## 1. **`NULL` mess up properties**

The nature of JSONB (and GeoJSON) is, that there is no specified properties list. However, the workaround we used above makes it that each feature has the same properties, but not values for all properties.
This results in some very weird export where value are mixed up and show on a different key.

| correct via db/vector tiles | wrong via export |
|--------|---|
| <img width="505" alt="image" src="https://github.com/user-attachments/assets/3acddf6a-8574-4617-bff7-f801d331aa25" /> [Link](https://radverkehrsatlas.de/regionen/bb-kampagne?map=13.4%2F52.744%2F12.7645&config=bv3m8i.5ount1.36q&f=10%7Cway%2F91664189%7C12.759837%7C52.736084%7C12.765234%7C52.749802&v=2) | <img width="483" alt="image" src="https://github.com/user-attachments/assets/fec6c899-dd2b-40ff-ad5f-b1c44181a0f2" /> |

**New Workaround:**
This can be fixed by changing the workaround to always add an empty string `''`:

```ts
      const tagKeys = tagKeyQuery
        .map(({ key }) => `COALESCE(tags->>'${key}', '') as "${key}"`)
        .join(',\n')
```

**New Issue:**
However, that has a new issue, which is that when I this file as a basis for a Mapbox Tileset, I cannot use Vector Tile Conditions like `["has", "someProperty"]`, presumably because the property is present as an empty string. I could work around this but since other data sources like our vector tile data generated by Martin don't have the same behavior I cannot switch the data sources without changing the conditions.

It would be better, if those values where an actual `NULL` in the data.

## 2. Numbers become strings

The screenshot above shows that `length: 3434.33` became a string. This causes the same issue described above with working with the Maplibre Style Spec.

**Try 1:**

I tried working around this with…
```ts
      return shouldCastToNumber
        ? `              COALESCE(NULLIF(${columnType}->>'${key}', NULL)::numeric, NULL) as "${key}"`
        : `              COALESCE(${columnType}->>'${key}', '') as "${key}"`
```

Which fails with …
> Raw query failed. Code: `XX000`. Message: `ERROR: flatgeobuf: get_column_type: '1700' column type not supported`

**Try 2:**

When we use `::float8`, the export work. BUT export that have number and string values fail in Placemark Play and show up empty in QGIS.

```ts
      return shouldCastToNumber
        ? `              COALESCE(NULLIF(${columnType}->>'${key}', NULL)::float8, NULL) as "${key}"`
        : `              COALESCE(${columnType}->>'${key}', '') as "${key}"`
```

Error in https://play.placemark.io/:
> <img width="499" alt="image" src="https://github.com/user-attachments/assets/fa3644c6-4bde-4114-a6d0-c50fad36595f" />

---

It looks like …
- …there is no way to preserve the number type from JsonB
- …there is no way to preserve the `NULL` values
